### ISSUE TITLE
Is it possible to merge Octopus environment variables with BTDF environment settings?

### ISSUE THREAD

**Issue by eloekset** [BTDF/DeploymentFramework#451]
One advantage of using Octopus deploy is to let the owners of each environment provide settings without rebuilding the BTDF project. Is this possible somehow? Which approach should I take to implement this?

I've created a PR in my own PowerShell script repository over here: https://github.com/eloekset/publish-btdfBiztalkApplication/pull/6/files

My first attempt was to have Octopus modify the generated BTDF environment settings XML file, but these files are not yet generated when the PowerShell script runs in Octopus. In the PR I've made a comment about this, and whether it's possible to pass `PropsFromEnvSetting` into the MSBuild target to override settings that will be used.

**Comment by billchesnut** [comment]
I create a column in the SettingsFileGenerator.xml, like this, pass the file created to Octopus and use it xml file replacement.

[cid:image001.png@01D42F57.A6D20930]

Bill Chesnut
Azure MVP (BizTalk & Integration)
Melbourne, Australia

**Comment by eloekset** [comment]
That was my first attempt also. I haven't had time to investigate the `PropsFromEnvSetting` parameter yet, but if it doesn't work as I hope, I'll have to split my Octopus process into more detailed steps to be able to produce the XML settings file for the environment and replace its values before the rest of the deployment steps run.

**Comment by billchesnut** [comment]
I am using ifdef based off variables from the PropsFromEnvSettings

From Deployment.btdfproj:

  <ItemGroup>
    <PropsFromEnvSettings Include="SsoAppUserGroup;SsoAppAdminGroup;BTSACCOUNT;ServerName;SCLEnvironment" />
  </ItemGroup>

In PortBindingsMaster.xml

      

What are you trying to do with PropsFromEnvSettings

Bill Chesnut
Azure MVP (BizTalk & Integration)
Melbourne, Australia

**Comment by eloekset** [comment]
I found this comment and code in the BizTalkDeploymentFramework.targets file:
```
  <!-- A list of named MSBuild properties to be created based on values pulled from the settings spreadsheet.
      The name of the property specified here must exactly match the name specified in a row in the settings spreadsheet. -->
  <!--
  DO NOT UNCOMMENT. This is just to demonstrate the usage of the ItemGroup. Should only be included in a .btdfproj.
  <ItemGroup>
    <PropsFromEnvSettings Include="ssoAppUserGroup;ssoAppAdminGroup;BAMViewsAndAccounts" />
  </ItemGroup>
  -->
```

So my question is whether it should be possible to just pass in key value pairs for environment settings  as a property to the MSBuild task, and then expect these values to override whatever is produced from the SettingsFileGenerator.xml. If that's how it is mean to work, I could just define all environment settings as variables in Octopus, and leave them blank in the SettingsFileGenerator.xml to fully rely on Octopus variables for each environment.

**Comment by billchesnut** [comment]
Sorry, never tried that, I use Octopus to replace the values in the xml file that comes out of the settingfilegenerator.xml for the Octopus environment (and Octopus puts in the variables for TEST, UAT and PROD)

Which looks like this, before Octopus replaces the #{}, with the values from its environment.

[cid:image001.png@01D42F5C.44CCDBC0]

Bill Chesnut
Azure MVP (BizTalk & Integration)
Melbourne, Australia

**Comment by eloekset** [comment]
Finally I've got time to have another look at this. Now, as far as I can see, `PropsFromEnvSettings` is not meant to be used the way I tried in the PR that I linked to in my initial comment. It's rather used to replace msbuild properties with values from the **SettingsFileGenerator.xml** file, almost the opposite of what I tried to do.

So I'll split my Octopus process into more detailed steps to have the **EnvironmentSettingsExporter.exe** do it's job the usual BTFD way, and then maybe follow @billchesnut tip to import the output xml file into Octopus for xml transformation.

Thanks for your answers, @billchesnut!
